pinot-core/src/main/java/com/linkedin/pinot/core/operator/query/DictionaryBasedAggregationOperator.java

```diff
@@ -0,0 +1,115 @@
+
+import com.linkedin.pinot.common.segment.SegmentMetadata;
+
+import com.linkedin.pinot.core.operator.BaseOperator;
+import com.linkedin.pinot.core.operator.ExecutionStatistics;
+import com.linkedin.pinot.core.operator.blocks.IntermediateResultsBlock;
+import com.linkedin.pinot.core.query.aggregation.AggregationFunctionContext;
+import com.linkedin.pinot.core.query.aggregation.AggregationResultHolder;
+import com.linkedin.pinot.core.query.aggregation.DoubleAggregationResultHolder;
+import com.linkedin.pinot.core.query.aggregation.ObjectAggregationResultHolder;
+import com.linkedin.pinot.core.query.aggregation.function.AggregationFunction;
+import com.linkedin.pinot.core.query.aggregation.function.AggregationFunctionFactory;
+import com.linkedin.pinot.core.query.aggregation.function.customobject.MinMaxRangePair;
+import com.linkedin.pinot.core.segment.index.readers.Dictionary;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Aggregation operator that utilizes metadata for serving aggregation queries.
+ */
+public class DictionaryBasedAggregationOperator extends BaseOperator<IntermediateResultsBlock> {
+  private static final String OPERATOR_NAME = "DictionaryBasedAggregationOperator";
+
+  private final AggregationFunctionContext[] _aggregationFunctionContexts;
+  private final Map<String, Dictionary> _dictionaryMap;
+  private final SegmentMetadata _segmentMetadata;
+  private ExecutionStatistics _executionStatistics;
+
+  /**
+   * Constructor for the class.
+   *
+   * @param aggregationFunctionContexts Aggregation function contexts.
+   * @param segmentMetadata Segment metadata.
+   * @param dictionaryMap Map of column to its dictionary.
+   */
+  public DictionaryBasedAggregationOperator(AggregationFunctionContext[] aggregationFunctionContexts,
+      SegmentMetadata segmentMetadata, Map<String, Dictionary> dictionaryMap) {
+    _aggregationFunctionContexts = aggregationFunctionContexts;
+    _dictionaryMap = dictionaryMap;
+    _segmentMetadata = segmentMetadata;
+  }
+
+  @Override
+  protected IntermediateResultsBlock getNextBlock() {
+    int numAggregationFunctions = _aggregationFunctionContexts.length;
+    List<Object> aggregationResults = new ArrayList<>(numAggregationFunctions);
+
+    for (AggregationFunctionContext aggregationFunctionContext : _aggregationFunctionContexts) {
+      AggregationFunction function = aggregationFunctionContext.getAggregationFunction();
+      AggregationFunctionFactory.AggregationFunctionType functionType =
+          AggregationFunctionFactory.AggregationFunctionType.valueOf(function.getName().toUpperCase());
+      String column = aggregationFunctionContext.getAggregationColumns()[0];
+      Dictionary dictionary = _dictionaryMap.get(column);
+      AggregationResultHolder resultHolder;
+      switch (functionType) {
+
+        case MAX:
+          resultHolder = new DoubleAggregationResultHolder(dictionary.getDoubleValue(dictionary.length() - 1));
```

Comment:
Will this work with star-tree where dictionary is modified to include the ALL node? If not, then either we should not select this operator in the plan phase, or handle it here (prefer latter if it can be cleanly achieved).